File: CS/Unbound Columns/MainViewModel.cs
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.ComponentModel;
using System.Linq;
using System.Runtime.CompilerServices;
using System.Windows.Input;
using DevExpress.Data;
using DevExpress.Mvvm;
using DevExpress.Mvvm.Native;
using DevExpress.Mvvm.Xpf;

namespace Unbound_Columns {
    public class MainViewModel : ViewModelBase {
        public ObservableCollection<Column> Columns { get => GetProperty(() => Columns); set => SetProperty(() => Columns, value); }
        public ObservableCollection<Item> Items { get; }

        public MainViewModel() {
            Items = new ObservableCollection<Item>(Enumerable.Range(0, 10).Select(i => new Item { Id = i }));
            Items.ForEach(x => {
                x["Name"] = $"Item {Items.IndexOf(x)}";
                x["CreatedAt"] = DateTime.Now.AddDays(Items.IndexOf(x));
            });

            Columns = new ObservableCollection<Column> {
                new Column() { FieldName = "Id" },
                new Column() { FieldName = "Name", UnboundType = UnboundColumnType.String },
                new Column() { FieldName = "CreatedAt", UnboundType = UnboundColumnType.DateTime }
            };

            OnCustomUnboundColumnDataCommand = new DelegateCommand<UnboundColumnRowArgs>(OnCustomUnboundColumnData);
            AddColumnCommand = new DelegateCommand(AddColumn);
        }

        public void OnCustomUnboundColumnData(UnboundColumnRowArgs e) {
            if (e.IsGetData) e.Value = Items[e.SourceIndex][e.FieldName];

            if (e.IsSetData) Items[e.SourceIndex][e.FieldName] = e.Value;
        }

        public void AddColumn() {
            Columns.Add(new Column { FieldName = $"Value {Columns.Count}", UnboundType = UnboundColumnType.Integer });
        }

        public ICommand OnCustomUnboundColumnDataCommand { get; }
        public ICommand AddColumnCommand { get; }
    }

    public class Item : BindableBase {
        private readonly Dictionary<string, object> _customFieldValues = new Dictionary<string, object>();
        public int Id { get => GetProperty(() => Id); set => SetProperty(() => Id, value); }

        public object this[string fieldName] {
            get {
                _customFieldValues.TryGetValue(fieldName, out var value);
                return value;
            }

            set {
                _customFieldValues[fieldName] = value;
                RaisePropertyChanged(fieldName);
            }
        }
    }

    public class Column : BindableBase {
        public string FieldName { get => GetProperty(() => FieldName); set => SetProperty(() => FieldName, value); }
        public UnboundColumnType? UnboundType { get => GetProperty(() => UnboundType); set => SetProperty(() => UnboundType, value); }
    }
}

File: VB/Unbound_Columns/MainViewModel.vb
Imports System.Collections.Generic
Imports System.Collections.ObjectModel
Imports System.ComponentModel
Imports System.Linq
Imports System.Windows.Input
Imports DevExpress.Data
Imports DevExpress.Mvvm
Imports DevExpress.Mvvm.Native
Imports DevExpress.Mvvm.Xpf

Namespace Unbound_Columns

    Public Class MainViewModel
        Inherits ViewModelBase

        Public Property Columns As ObservableCollection(Of Column)
            Get
                Return GetProperty(Function() Me.Columns)
            End Get

            Set(ByVal value As ObservableCollection(Of Column))
                SetProperty(Function() Columns, value)
            End Set
        End Property

        Public ReadOnly Property Items As ObservableCollection(Of Item)

        Public Sub New()
            Items = New ObservableCollection(Of Item)(Enumerable.Range(0, 10).[Select](Function(i) New Item With {.Id = i}))
            Items.ForEach(Sub(x)
                x("Name") = $"Item {Items.IndexOf(x)}"
                x("CreatedAt") = Date.Now.AddDays(Items.IndexOf(x))
            End Sub)
            Columns = New ObservableCollection(Of Column) From {New Column() With {.FieldName = "Id"}, New Column() With {.FieldName = "Name", .UnboundType = UnboundColumnType.String}, New Column() With {.FieldName = "CreatedAt", .UnboundType = UnboundColumnType.DateTime}}
            OnCustomUnboundColumnDataCommand = New DelegateCommand(Of UnboundColumnRowArgs)(AddressOf OnCustomUnboundColumnData)
            AddColumnCommand = New DelegateCommand(AddressOf AddColumn)
        End Sub

        Public Sub OnCustomUnboundColumnData(ByVal e As UnboundColumnRowArgs)
            If e.IsGetData Then e.Value = Items(e.SourceIndex)(e.FieldName)
            If e.IsSetData Then Items(e.SourceIndex)(e.FieldName) = e.Value
        End Sub

        Public Sub AddColumn()
            Columns.Add(New Column With {.FieldName = $"Value {Columns.Count}", .UnboundType = UnboundColumnType.Integer})
        End Sub

        Public ReadOnly Property OnCustomUnboundColumnDataCommand As ICommand

        Public ReadOnly Property AddColumnCommand As ICommand
    End Class

    Public Class Item
        Inherits BindableBase

        Private ReadOnly _customFieldValues As Dictionary(Of String, Object) = New Dictionary(Of String, Object)()

        Public Property Id As Integer
            Get
                Return GetProperty(Function() Me.Id)
            End Get

            Set(ByVal value As Integer)
                SetProperty(Function() Id, value)
            End Set
        End Property

        Default Public Property Item(ByVal fieldName As String) As Object
            Get
                Dim value As Object = Nothing
                _customFieldValues.TryGetValue(fieldName, value)
                Return value
            End Get

            Set(ByVal value As Object)
                _customFieldValues(fieldName) = value
                RaisePropertyChanged(fieldName)
            End Set
        End Property
    End Class

    Public Class Column
        Inherits BindableBase

        Public Property FieldName As String
            Get
                Return GetProperty(Function() Me.FieldName)
            End Get

            Set(ByVal value As String)
                SetProperty(Function() FieldName, value)
            End Set
        End Property

        Public Property UnboundType As UnboundColumnType?
            Get
                Return GetProperty(Function() Me.UnboundType)
            End Get

            Set(ByVal value As UnboundColumnType?)
                SetProperty(Function() UnboundType, value)
            End Set
        End Property
    End Class
End Namespace
